src/tcache.c

```diff
@@ -424,28 +427,37 @@ tcache_stats_merge(tsdn_t *tsdn, tcache_t *tcache, arena_t *arena) {
 
 bool
 tcaches_create(tsd_t *tsd, unsigned *r_ind) {
+	witness_assert_depth(tsd_tsdn(tsd), 0);
+
+	bool err;
 	arena_t *arena;
 	tcache_t *tcache;
 	tcaches_t *elm;
 
+	malloc_mutex_lock(tsd_tsdn(tsd), &tcaches_mtx);
+
 	if (tcaches == NULL) {
 		tcaches = base_alloc(tsd_tsdn(tsd), b0get(), sizeof(tcache_t *)
 		    * (MALLOCX_TCACHE_MAX+1), CACHELINE);
 		if (tcaches == NULL) {
-			return true;
+			err = true;
+			goto label_return;
 		}
 	}
 
 	if (tcaches_avail == NULL && tcaches_past > MALLOCX_TCACHE_MAX) {
-		return true;
+		err = true;
+		goto label_return;
 	}
 	arena = arena_ichoose(tsd, NULL);
```

Comment:
I think this can trigger arena creation? If so maybe consider releasing tcaches_mtx before this. Looks like if we move the { tcaches_past++; }  branch to before this, then the code after doesn't need to be protected.